server.port lookup fails for complex values

### Describe the bug

I have a spring-boot app with the following application.yml (top only):

```yaml
 server:
  port: ${APP_SERVER_PORT:8080}
```

And I am using the spring boot generator which tries to find the HTTP port from the config: https://eclipse.dev/jkube/docs/kubernetes-maven-plugin/#generator-spring-boot 

When running

```
./mvnw k8s:resource
```

And a jkube configuration like the following:

```xml
            <plugin>
                <groupId>org.eclipse.jkube</groupId>
                <artifactId>kubernetes-maven-plugin</artifactId>
                <version>${jkube.version}</version>
                <configuration>
                    <generator>
                        <config>
                            <spring-boot>
                                <name>docker.io/library/${project.artifactId}:${project.version}</name>
                            </spring-boot>
                        </config>
                    </generator>
                </configuration>
            </plugin>
```


With maven 3.9.4 (outdated values in the form ;) ), I get the following error in the output

```
[INFO] --- k8s:1.15.0:resource (default-cli) @ microservice-app ---
[INFO] k8s: Using Dockerfile: /Volumes/Workspace/b2c/template-springboot-microservice/microservice-app/Dockerfile
[INFO] k8s: Using Docker Context Directory: /Volumes/Workspace/b2c/template-springboot-microservice/microservice-app
[INFO] k8s: Using resource templates from /Volumes/Workspace/b2c/template-springboot-microservice/microservice-app/src/main/jkube
[INFO] k8s: jkube-controller: Adding a default Deployment
[ERROR] k8s: jkube-healthcheck-spring-boot: Error while reading the spring-boot configuration: java.lang.NumberFormatException: For input string: "${APP_SERVER_PORT:8080}"
[ERROR] k8s: jkube-healthcheck-spring-boot: Error while reading the spring-boot configuration: java.lang.NumberFormatException: For input string: "${APP_SERVER_PORT:8080}"
[INFO] k8s: jkube-revision-history: Adding revision history limit to 2
[INFO] k8s: validating /Volumes/Workspace/b2c/template-springboot-microservice/microservice-app/target/classes/META-INF/jkube/kubernetes/microservice-app-deployment.yml resource
```

Thanks!



### Eclipse JKube version

1.15.0

### Component

Kubernetes Maven Plugin

### Apache Maven version

other (please specify in additional context)

### Gradle version

None

### Steps to reproduce

1. Create a standard maven spring-boot project with a variable as the `server.port` value in the `application.yml` file
2. Configure the jkube maven with the spring boot generator
3. Run `./mvnw k8s:resource`


### Expected behavior

I would expect:
* best option: the detection code looks into the variable and picks the default value (8080 in this case)
* worst option: the detection code fails gracefully and picks the default

### Runtime

other (please specify in additional context)

### Kubernetes API Server version

1.25.3

### Environment

macOS

### Eclipse JKube Logs

```shell
[INFO] --- k8s:1.15.0:resource (default-cli) @ microservice-app ---
[INFO] k8s: Using Dockerfile: /Volumes/Workspace/b2c/template-springboot-microservice/microservice-app/Dockerfile
[INFO] k8s: Using Docker Context Directory: /Volumes/Workspace/b2c/template-springboot-microservice/microservice-app
[INFO] k8s: Using resource templates from /Volumes/Workspace/b2c/template-springboot-microservice/microservice-app/src/main/jkube
[INFO] k8s: jkube-controller: Adding a default Deployment
[ERROR] k8s: jkube-healthcheck-spring-boot: Error while reading the spring-boot configuration: java.lang.NumberFormatException: For input string: "${APP_SERVER_PORT:8080}"
[ERROR] k8s: jkube-healthcheck-spring-boot: Error while reading the spring-boot configuration: java.lang.NumberFormatException: For input string: "${APP_SERVER_PORT:8080}"
[INFO] k8s: jkube-revision-history: Adding revision history limit to 2
[INFO] k8s: validating /Volumes/Workspace/b2c/template-springboot-microservice/microservice-app/target/classes/META-INF/jkube/kubernetes/microservice-app-deployment.yml resource
```


### Sample Reproducer Project

_No response_

### Additional context

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -23,6 +23,7 @@ Usage:
 ### 1.16-SNAPSHOT
 * Fix #1690: Base images based on ubi9
 * Fix #2390: support for all missing Chart.yaml fields
+* Fix #2444: Add support for Spring Boot application properties placeholders
 
 ### 1.15.0 (2023-11-10)
 * Fix #2138: Support for Spring Boot Native Image

--- a/gradle-plugin/it/src/it/spring-boot-properties/build.gradle
+++ b/gradle-plugin/it/src/it/spring-boot-properties/build.gradle
@@ -1,0 +1,39 @@
+/*
+ * Copyright (c) 2019 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at:
+ *
+ *     https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+plugins {
+  id 'org.eclipse.jkube.kubernetes' version "${jKubeVersion}"
+  id 'org.eclipse.jkube.openshift' version "${jKubeVersion}"
+  id 'org.springframework.boot' version '2.7.17'
+  id 'io.spring.dependency-management' version '1.0.15.RELEASE'
+  id 'java'
+}
+
+group = 'org.eclipse.jkube.integration.tests.gradle'
+version = '0.0.1-SNAPSHOT'
+
+repositories {
+  mavenCentral()
+}
+
+dependencies {
+  implementation 'org.springframework.boot:spring-boot-starter-web'
+  implementation 'org.springframework.boot:spring-boot-starter-actuator'
+}
+
+def extensionConfig = {
+  offline = true
+}
+
+kubernetes(extensionConfig)
+openshift(extensionConfig)

--- a/gradle-plugin/it/src/it/spring-boot-properties/expected/kubernetes.yml
+++ b/gradle-plugin/it/src/it/spring-boot-properties/expected/kubernetes.yml
@@ -1,0 +1,89 @@
+#
+# Copyright (c) 2019 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at:
+#
+#     https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#   Red Hat, Inc. - initial API and implementation
+#
+
+---
+apiVersion: v1
+kind: List
+items:
+- apiVersion: v1
+  kind: Service
+  metadata:
+    name: spring-boot-properties
+  spec:
+    ports:
+    - name: irdmi
+      port: 8000
+      protocol: TCP
+      targetPort: 8000
+    selector:
+      app: spring-boot-properties
+      provider: jkube
+      group: org.eclipse.jkube.integration.tests.gradle
+- apiVersion: apps/v1
+  kind: Deployment
+  metadata:
+    labels:
+      provider: jkube
+      app: spring-boot-properties
+      version: "@ignore@"
+      group: org.eclipse.jkube.integration.tests.gradle
+    name: spring-boot-properties
+  spec:
+    replicas: 1
+    selector:
+      matchLabels:
+        app: spring-boot-properties
+        provider: jkube
+        group: org.eclipse.jkube.integration.tests.gradle
+    template:
+      metadata:
+        labels:
+          provider: jkube
+          app: spring-boot-properties
+          version: "@ignore@"
+          group: org.eclipse.jkube.integration.tests.gradle
+      spec:
+        containers:
+        - env:
+          - name: KUBERNETES_NAMESPACE
+            valueFrom:
+              fieldRef:
+                fieldPath: metadata.namespace
+          image: "@matches('gradle/spring-boot-properties:.*$')@"
+          imagePullPolicy: IfNotPresent
+          livenessProbe:
+            httpGet:
+              path: /actuator/health
+              port: 8008
+              scheme: HTTP
+            initialDelaySeconds: 180
+          name: spring-boot
+          ports:
+          - containerPort: 8000
+            name: irdmi
+            protocol: TCP
+          - containerPort: 9779
+            name: prometheus
+            protocol: TCP
+          - containerPort: 8778
+            name: jolokia
+            protocol: TCP
+          readinessProbe:
+            httpGet:
+              path: /actuator/health
+              port: 8008
+              scheme: HTTP
+            initialDelaySeconds: 10
+          securityContext:
+            privileged: false

--- a/gradle-plugin/it/src/it/spring-boot-properties/expected/openshift.yml
+++ b/gradle-plugin/it/src/it/spring-boot-properties/expected/openshift.yml
@@ -1,0 +1,102 @@
+#
+# Copyright (c) 2019 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at:
+#
+#     https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#   Red Hat, Inc. - initial API and implementation
+#
+
+---
+apiVersion: v1
+kind: List
+items:
+- apiVersion: v1
+  kind: Service
+  metadata:
+    name: spring-boot-properties
+  spec:
+    ports:
+    - name: irdmi
+      port: 8000
+      protocol: TCP
+      targetPort: 8000
+    selector:
+      app: spring-boot-properties
+      provider: jkube
+      group: org.eclipse.jkube.integration.tests.gradle
+- apiVersion: apps.openshift.io/v1
+  kind: DeploymentConfig
+  metadata:
+    labels:
+      provider: jkube
+      app: spring-boot-properties
+      version: "@ignore@"
+      group: org.eclipse.jkube.integration.tests.gradle
+    name: spring-boot-properties
+  spec:
+    replicas: 1
+    selector:
+      app: spring-boot-properties
+      provider: jkube
+      group: org.eclipse.jkube.integration.tests.gradle
+    strategy:
+      rollingParams:
+        timeoutSeconds: 3600
+      type: Rolling
+    template:
+      metadata:
+        labels:
+          provider: jkube
+          app: spring-boot-properties
+          version: "@ignore@"
+          group: org.eclipse.jkube.integration.tests.gradle
+      spec:
+        containers:
+        - env:
+          - name: KUBERNETES_NAMESPACE
+            valueFrom:
+              fieldRef:
+                fieldPath: metadata.namespace
+          image: spring-boot-properties:latest
+          imagePullPolicy: IfNotPresent
+          livenessProbe:
+            httpGet:
+              path: /actuator/health
+              port: 8008
+              scheme: HTTP
+            initialDelaySeconds: 180
+          name: spring-boot
+          ports:
+          - containerPort: 8000
+            name: irdmi
+            protocol: TCP
+          - containerPort: 9779
+            name: prometheus
+            protocol: TCP
+          - containerPort: 8778
+            name: jolokia
+            protocol: TCP
+          readinessProbe:
+            httpGet:
+              path: /actuator/health
+              port: 8008
+              scheme: HTTP
+            initialDelaySeconds: 10
+          securityContext:
+            privileged: false
+    triggers:
+    - type: ConfigChange
+    - imageChangeParams:
+        automatic: true
+        containerNames:
+        - spring-boot
+        from:
+          kind: ImageStreamTag
+          name: spring-boot-properties:latest
+      type: ImageChange

--- a/gradle-plugin/it/src/it/spring-boot-properties/src/main/java/org/eclipse/jkube/it/gradle/spring/boot/Application.java
+++ b/gradle-plugin/it/src/it/spring-boot-properties/src/main/java/org/eclipse/jkube/it/gradle/spring/boot/Application.java
@@ -1,0 +1,26 @@
+/*
+ * Copyright (c) 2019 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at:
+ *
+ *     https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.jkube.it.gradle.spring.boot;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+public class Application {
+
+  public static void main(String[] args) {
+        SpringApplication.run(Application.class, args);
+    }
+
+}

--- a/gradle-plugin/it/src/it/spring-boot-properties/src/main/resources/application.properties
+++ b/gradle-plugin/it/src/it/spring-boot-properties/src/main/resources/application.properties
@@ -1,0 +1,17 @@
+#
+# Copyright (c) 2019 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at:
+#
+#     https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#   Red Hat, Inc. - initial API and implementation
+#
+
+default.server.port=8000
+server.port=${APP_SERVER_PORT:${default.server.port}}
+management.server.port=8008

--- a/gradle-plugin/it/src/test/java/org/eclipse/jkube/gradle/plugin/tests/SpringBootPropertiesIT.java
+++ b/gradle-plugin/it/src/test/java/org/eclipse/jkube/gradle/plugin/tests/SpringBootPropertiesIT.java
@@ -1,0 +1,67 @@
+/*
+ * Copyright (c) 2019 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at:
+ *
+ *     https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.jkube.gradle.plugin.tests;
+
+import net.minidev.json.parser.ParseException;
+import org.eclipse.jkube.kit.common.ResourceVerify;
+import org.gradle.testkit.runner.BuildResult;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import java.io.IOException;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+class SpringBootPropertiesIT {
+
+  @RegisterExtension
+  public final ITGradleRunnerExtension gradleRunner = new ITGradleRunnerExtension();
+
+  @Test
+  void k8sResource_whenRun_generatesK8sManifestsWithProperties() throws IOException, ParseException {
+    // When
+    final BuildResult result = gradleRunner.withITProject("spring-boot-properties")
+      .withArguments("build", "k8sResource", "--stacktrace")
+      .build();
+    // Then
+    ResourceVerify.verifyResourceDescriptors(gradleRunner.resolveDefaultKubernetesResourceFile(),
+      gradleRunner.resolveFile("expected", "kubernetes.yml"));
+    assertThat(result).extracting(BuildResult::getOutput).asString()
+      .contains("jkube-controller: Adding a default Deployment")
+      .contains("jkube-service: Adding a default service")
+      .contains("jkube-healthcheck-spring-boot: Adding readiness probe on port 8008")
+      .contains("jkube-healthcheck-spring-boot: Adding liveness probe on port 8008")
+      .contains("jkube-service-discovery: Using first mentioned service port '8000' ")
+      .contains("jkube-revision-history: Adding revision history limit to 2");
+  }
+
+  @Test
+  void ocResource_whenRun_generatesOpenShiftManifestsWithProperties() throws IOException, ParseException {
+    // When
+    final BuildResult result = gradleRunner.withITProject("spring-boot-properties")
+      .withArguments("build", "ocResource", "--stacktrace")
+      .build();
+    // Then
+    ResourceVerify.verifyResourceDescriptors(gradleRunner.resolveDefaultOpenShiftResourceFile(),
+      gradleRunner.resolveFile("expected", "openshift.yml"));
+    assertThat(result).extracting(BuildResult::getOutput).asString()
+      .contains("jkube-controller: Adding a default Deployment")
+      .contains("jkube-service: Adding a default service")
+      .contains("jkube-openshift-deploymentconfig: Converting Deployment to DeploymentConfig")
+      .contains("jkube-healthcheck-spring-boot: Adding readiness probe on port 8008")
+      .contains("jkube-healthcheck-spring-boot: Adding liveness probe on port 8008")
+      .contains("jkube-service-discovery: Using first mentioned service port '8000' ")
+      .contains("jkube-revision-history: Adding revision history limit to 2");
+  }
+}

--- a/jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/SpringBootPropertyPlaceholderHelper.java
+++ b/jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/SpringBootPropertyPlaceholderHelper.java
@@ -1,0 +1,180 @@
+/*
+ * Copyright (c) 2019 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at:
+ *
+ *     https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.jkube.kit.common.util;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.Set;
+
+/**
+ * Ported/inspired from
+ * <a href="https://github.com/spring-projects/spring-framework/blob/7b14606d92bfe36fea048137e794cfff18e9e42a/spring-core/src/main/java/org/springframework/util/PropertyPlaceholderHelper.java">PropertyPlaceholderHelper</a>
+ */
+public class SpringBootPropertyPlaceholderHelper {
+
+  private static final Map<String, String> wellKnownSimplePrefixes = new HashMap<>(4);
+
+  private final String placeholderPrefix;
+
+  private final String placeholderSuffix;
+
+  private final String simplePrefix;
+
+  private final String valueSeparator;
+
+  private final boolean ignoreUnresolvablePlaceholders;
+
+  public SpringBootPropertyPlaceholderHelper(String placeholderPrefix, String placeholderSuffix) {
+    this(placeholderPrefix, placeholderSuffix, null, true);
+  }
+
+  public SpringBootPropertyPlaceholderHelper(String placeholderPrefix, String placeholderSuffix,
+                                             String valueSeparator, boolean ignoreUnresolvablePlaceholders) {
+
+    Objects.requireNonNull(placeholderPrefix, "'placeholderPrefix' must not be null");
+    Objects.requireNonNull(placeholderSuffix, "'placeholderSuffix' must not be null");
+    this.placeholderPrefix = placeholderPrefix;
+    this.placeholderSuffix = placeholderSuffix;
+    String simplePrefixForSuffix = wellKnownSimplePrefixes.get(this.placeholderSuffix);
+    if (simplePrefixForSuffix != null && this.placeholderPrefix.endsWith(simplePrefixForSuffix)) {
+      this.simplePrefix = simplePrefixForSuffix;
+    } else {
+      this.simplePrefix = this.placeholderPrefix;
+    }
+    this.valueSeparator = valueSeparator;
+    this.ignoreUnresolvablePlaceholders = ignoreUnresolvablePlaceholders;
+  }
+
+  public Properties replaceAllPlaceholders(Properties properties) {
+    Objects.requireNonNull(properties, "'properties' must not be null");
+    final Properties result = new Properties();
+    for (Map.Entry<Object, Object> entry : properties.entrySet()) {
+      result.put(entry.getKey(), replacePlaceholders((String) entry.getValue(), properties));
+    }
+    return result;
+  }
+
+  public String replacePlaceholders(String value, final Properties properties) {
+    Objects.requireNonNull(properties, "'properties' must not be null");
+    return replacePlaceholders(value, properties::getProperty);
+  }
+
+  public String replacePlaceholders(String value, PlaceholderResolver placeholderResolver) {
+    Objects.requireNonNull(value, "'value' must not be null");
+    return parseStringValue(value, placeholderResolver, null);
+  }
+
+  private String parseStringValue(
+    String value, PlaceholderResolver placeholderResolver,  Set<String> visitedPlaceholders) {
+
+    int startIndex = value.indexOf(this.placeholderPrefix);
+    if (startIndex == -1) {
+      return value;
+    }
+
+    StringBuilder result = new StringBuilder(value);
+    while (startIndex != -1) {
+      int endIndex = findPlaceholderEndIndex(result, startIndex);
+      if (endIndex != -1) {
+        String placeholder = result.substring(startIndex + this.placeholderPrefix.length(), endIndex);
+        String originalPlaceholder = placeholder;
+        if (visitedPlaceholders == null) {
+          visitedPlaceholders = new HashSet<>(4);
+        }
+        if (!visitedPlaceholders.add(originalPlaceholder)) {
+          throw new IllegalArgumentException(
+            "Circular placeholder reference '" + originalPlaceholder + "' in property definitions");
+        }
+        // Recursive invocation, parsing placeholders contained in the placeholder key.
+        placeholder = parseStringValue(placeholder, placeholderResolver, visitedPlaceholders);
+        // Now obtain the value for the fully resolved key...
+        String propVal = placeholderResolver.resolvePlaceholder(placeholder);
+        if (propVal == null && this.valueSeparator != null) {
+          int separatorIndex = placeholder.indexOf(this.valueSeparator);
+          if (separatorIndex != -1) {
+            String actualPlaceholder = placeholder.substring(0, separatorIndex);
+            String defaultValue = placeholder.substring(separatorIndex + this.valueSeparator.length());
+            propVal = placeholderResolver.resolvePlaceholder(actualPlaceholder);
+            if (propVal == null) {
+              propVal = defaultValue;
+            }
+          }
+        }
+        if (propVal != null) {
+          // Recursive invocation, parsing placeholders contained in the
+          // previously resolved placeholder value.
+          propVal = parseStringValue(propVal, placeholderResolver, visitedPlaceholders);
+          result.replace(startIndex, endIndex + this.placeholderSuffix.length(), propVal);
+          startIndex = result.indexOf(this.placeholderPrefix, startIndex + propVal.length());
+        }
+        else if (this.ignoreUnresolvablePlaceholders) {
+          // Proceed with unprocessed value.
+          startIndex = result.indexOf(this.placeholderPrefix, endIndex + this.placeholderSuffix.length());
+        }
+        else {
+          throw new IllegalArgumentException("Could not resolve placeholder '" +
+            placeholder + "'" + " in value \"" + value + "\"");
+        }
+        visitedPlaceholders.remove(originalPlaceholder);
+      }
+      else {
+        startIndex = -1;
+      }
+    }
+    return result.toString();
+  }
+
+  private int findPlaceholderEndIndex(CharSequence buf, int startIndex) {
+    int index = startIndex + this.placeholderPrefix.length();
+    int withinNestedPlaceholder = 0;
+    while (index < buf.length()) {
+      if (StringUtils.startsWith(buf.subSequence(index, buf.length()), this.placeholderSuffix)) {
+        if (withinNestedPlaceholder > 0) {
+          withinNestedPlaceholder--;
+          index = index + this.placeholderSuffix.length();
+        }
+        else {
+          return index;
+        }
+      }
+      else if (StringUtils.startsWith(buf.subSequence(index, buf.length()), this.simplePrefix)) {
+        withinNestedPlaceholder++;
+        index = index + this.simplePrefix.length();
+      }
+      else {
+        index++;
+      }
+    }
+    return -1;
+  }
+
+  /**
+   * Strategy interface used to resolve replacement values for placeholders contained in Strings.
+   */
+  @FunctionalInterface
+  public interface PlaceholderResolver {
+
+    /**
+     * Resolve the supplied placeholder name to the replacement value.
+     * @param placeholderName the name of the placeholder to resolve
+     * @return the replacement value, or {@code null} if no replacement is to be made
+     */
+    String resolvePlaceholder(String placeholderName);
+  }
+}

--- a/jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/SpringBootUtil.java
+++ b/jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/SpringBootUtil.java
@@ -41,6 +41,10 @@ public class SpringBootUtil {
     public static final String DEV_TOOLS_REMOTE_SECRET = "spring.devtools.remote.secret";
     public static final String DEV_TOOLS_REMOTE_SECRET_ENV = "SPRING_DEVTOOLS_REMOTE_SECRET";
 
+    private static final String PLACEHOLDER_PREFIX = "${";
+    private static final String PLACEHOLDER_SUFFIX = "}";
+    private static final String VALUE_SEPARATOR = ":";
+
     private SpringBootUtil() {}
 
     /**
@@ -68,7 +72,8 @@ public class SpringBootUtil {
 
         Properties props = YamlUtil.getPropertiesFromYamlResource(springActiveProfile, ymlResource);
         props.putAll(getPropertiesFromResource(propertiesResource));
-        return props;
+        return new SpringBootPropertyPlaceholderHelper(PLACEHOLDER_PREFIX, PLACEHOLDER_SUFFIX, VALUE_SEPARATOR, true)
+        .replaceAllPlaceholders(props);
     }
 
     /**

--- a/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/SpringBootPropertyPlaceholderHelperTest.java
+++ b/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/SpringBootPropertyPlaceholderHelperTest.java
@@ -1,0 +1,100 @@
+/*
+ * Copyright (c) 2019 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at:
+ *
+ *     https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.jkube.kit.common.util;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Properties;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.eclipse.jkube.kit.common.util.SpringBootUtilTest.createClassLoader;
+
+class SpringBootPropertyPlaceholderHelperTest {
+
+  private Properties result;
+
+  @BeforeEach
+  void setUp(@TempDir File tempDir) throws IOException {
+    final URLClassLoader classLoader = createClassLoader(tempDir, "/util/springboot/spring-boot-complex-application.properties");
+    result =  SpringBootUtil.getSpringBootApplicationProperties(classLoader);
+  }
+
+  @Test
+  void resolvesStandard() {
+    assertThat(result).containsEntry("custom.property.name", "my-application");
+  }
+
+  @Test
+  void resolvesSimpleDefaultValue() {
+    assertThat(result).containsEntry("server.port", "8082");
+  }
+
+  @Test
+  void resolvesStackedDefaultValue() {
+    assertThat(result).containsEntry("custom.property.db-host", "example.com");
+  }
+
+  @Test
+  void resolvesInterleavedWithStackedDefaultValue() {
+    assertThat(result).containsEntry("app.description", "An application with the name my-application");
+  }
+
+  @Test
+  void resolvesMultipleInterleavedWithStackedDefaultValue() {
+    assertThat(result).containsEntry("custom.property.url", "https://example.com");
+  }
+
+  @Test
+  void resolvesMultipleInterleavedWithDeepStackedDefaultValue() {
+    assertThat(result).containsEntry("spring.datasource.url", "jdbc:postgresql://example.com/production");
+  }
+
+  @Test
+  void resolvesNestedWithStackedDefaultValue() {
+    assertThat(result).containsEntry("custom.property.nested", "dev.example.com");
+  }
+  @Test
+  void resolvesDefaultNestedValue() {
+    assertThat(result).containsEntry("custom.property.nested.default", "dev");
+  }
+
+  @Test
+  void failsWithCircular(@TempDir Path tempDir) throws IOException {
+    final Path applicationProperties = tempDir.resolve("application.properties");
+    Files.write(applicationProperties, Collections.singletonList("custom.property.circular=${custom.property.circular}"));
+    try (URLClassLoader classLoader = ClassUtil.createClassLoader(Arrays.asList(
+      tempDir.toFile().getAbsolutePath(),
+      applicationProperties.toFile().getAbsolutePath()
+    ))) {
+      assertThatThrownBy(() -> SpringBootUtil.getSpringBootApplicationProperties(classLoader))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("Circular placeholder reference 'custom.property.circular' in property definitions");
+    }
+  }
+
+  @Test
+  void ignoresUnresolvedValues() {
+    assertThat(result).containsEntry("custom.property.unresolved", "${unresolved}");
+  }
+}

--- a/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/SpringBootUtilTest.java
+++ b/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/SpringBootUtilTest.java
@@ -39,8 +39,6 @@ import java.util.jar.Manifest;
 
 import static org.assertj.core.api.Assertions.assertThatIllegalStateException;
 import static org.junit.jupiter.api.Assertions.assertThrows;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.when;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.entry;
 
@@ -50,7 +48,9 @@ class SpringBootUtilTest {
 
     @BeforeEach
     public void setUp() {
-        mavenProject = mock(JavaProject.class);
+        mavenProject = JavaProject.builder()
+          .properties(new Properties())
+          .build();
     }
 
     @Test
@@ -67,10 +67,11 @@ class SpringBootUtilTest {
     }
 
     @Test
-    void testGetSpringBootVersion() {
+    void getSpringBootVersion() {
         //Given
-        Dependency p = Dependency.builder().groupId("org.springframework.boot").version("1.6.3").build();
-        when(mavenProject.getDependencies()).thenReturn(Collections.singletonList(p));
+        mavenProject = mavenProject.toBuilder()
+          .dependency(Dependency.builder().groupId("org.springframework.boot").version("1.6.3").build())
+          .build();
 
         //when
         Optional<String> result = SpringBootUtil.getSpringBootVersion(mavenProject);
@@ -80,11 +81,9 @@ class SpringBootUtilTest {
     }
 
     @Test
-    void testGetSpringBootActiveProfileWhenNotNull() {
+    void getSpringBootActiveProfileWhenNotNull() {
         //Given
-        Properties p = new Properties();
-        p.put("spring.profiles.active","spring-boot");
-        when(mavenProject.getProperties()).thenReturn(p);
+        mavenProject.getProperties().put("spring.profiles.active","spring-boot");
 
         // When
         String result = SpringBootUtil.getSpringBootActiveProfile(mavenProject);
@@ -420,7 +419,7 @@ class SpringBootUtilTest {
         assertThat(nativeArtifactFound).hasName("sample");
     }
 
-    private URLClassLoader createClassLoader(File temporaryFolder, String resource) throws IOException {
+    static URLClassLoader createClassLoader(File temporaryFolder, String resource) throws IOException {
         File applicationProp =  new File(Objects.requireNonNull(SpringBootUtilTest.class.getResource(resource)).getPath());
         File classesInTarget = new File(new File(temporaryFolder, "target"), "classes");
         FileUtil.createDirectory(classesInTarget);

--- a/jkube-kit/common/src/test/resources/util/springboot/spring-boot-complex-application.properties
+++ b/jkube-kit/common/src/test/resources/util/springboot/spring-boot-complex-application.properties
@@ -1,0 +1,29 @@
+#
+# Copyright (c) 2019 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at:
+#
+#     https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#   Red Hat, Inc. - initial API and implementation
+#
+
+custom.property.name=my-application
+custom.property.db=${DB:production}
+custom.property.db-host=${custom.property.domain}
+custom.property.domain=${HOSTNAME:example.com}
+custom.property.nested=${custom.property.nested.domain.${custom.property.nested.env}}
+custom.property.nested.default=${SOME_VAR:${custom.property.nested.env}}
+custom.property.nested.domain.dev=dev.example.com
+custom.property.nested.env=dev
+custom.property.protocol=https
+custom.property.unresolved=${unresolved}
+custom.property.url=${custom.property.protocol}://${custom.property.domain}
+app.name=${custom.property.name}
+app.description=An application with the name ${app.name}
+spring.datasource.url=jdbc:postgresql://${custom.property.db-host}/${custom.property.db}
+server.port=${APP_SERVER_PORT:8082}

